File: paseto-base/src/main/java/net/aholbrook/paseto/verification/rules/ForAudience.java
package net.aholbrook.paseto.verification.rules;

import net.aholbrook.paseto.Token;
import net.aholbrook.paseto.exception.verification.rules.IncorrectAudienceException;
import net.aholbrook.paseto.exception.verification.rules.MissingClaimException;
import net.aholbrook.paseto.util.StringUtils;
import net.aholbrook.paseto.verification.PasetoVerificationContext;

public class ForAudience implements Rule {
	public final static String NAME = "FOR_AUDIENCE";

	private final String audience;

	/**
	 * Verifies that the token Audience (aud) claim matches the given value.
	 * @param audience The expected audience for the token.
	 */
	public ForAudience(String audience) {
		this.audience = StringUtils.ntes(audience);
	}

	@Override
	public String name() {
		return NAME;
	}

	@Override
	public void check(Token token, PasetoVerificationContext context) {
		if (StringUtils.isEmpty(token.getAudience())) {
			throw new MissingClaimException(Token.CLAIM_AUDIENCE, NAME, token);
		}

		if (!audience.equals(token.getAudience())) {
			throw new IncorrectAudienceException(audience, token.getAudience(), NAME, token);
		}
	}
}

File: paseto/src/main/kotlin/net/aholbrook/paseto/rules/ForAudience.kt
package net.aholbrook.paseto.rules

import net.aholbrook.paseto.Token
import net.aholbrook.paseto.crypto.constantTimeEquals
import net.aholbrook.paseto.exception.IncorrectAudienceException
import net.aholbrook.paseto.exception.MissingClaimException

/**
 * Verifies that the token Audience (aud) claim matches the given value.
 *
 * @param audience The expected audience for the token.
 */
data class ForAudience(val audience: String) : Rule {
    override operator fun invoke(token: Token, mode: Rule.Mode, currentResults: Map<Rule, RuleResult>) {
        if (token.audience.isNullOrEmpty()) {
            throw MissingClaimException("aud", token)
        }

        if (!token.audience.constantTimeEquals(audience)) {
            throw IncorrectAudienceException(audience, token.audience, token)
        }
    }
}
